packages/marshal/src/marshal.js

```diff
@@ -216,28 +287,84 @@ export const makeMarshal = (
       return val;
     };
 
-    const fullRevive = makeDecodeFromCapData({
+    const decodeRemotableFromSmallcaps = stringEncoding => {
+      assert(stringEncoding.startsWith('$'));
+      // slots: $slotIndex.iface or $slotIndex
+      const i = stringEncoding.indexOf('.');
+      const index = Number(stringEncoding.slice(1, i));
+      // assert(i > 0, X`${rawTree} is not a valid slot encoding`);
+      // i < 0 means there was no iface included
+      const iface = i < 0 ? undefined : stringEncoding.slice(i + 1);
+      return unserializeSlot(index, iface);
+    };
+
+    const decodePromiseFromSmallcaps = stringEncoding => {
+      assert(stringEncoding.startsWith('&'));
+      // slots: $slotIndex.iface or $slotIndex
+      const i = stringEncoding.indexOf('.');
+      const index = Number(stringEncoding.slice(1, i));
+      // assert(i > 0, X`${rawTree} is not a valid slot encoding`);
+      // i < 0 means there was no iface included
+      const iface = i < 0 ? undefined : stringEncoding.slice(i + 1);
+      return unserializeSlot(index, iface);
+    };
+
+    const decodeErrorFromSmallcaps = encoding => {
+      const { '#error': message, name, ...rest } = encoding;
+      // Not the most elegant way to reuse code. TODO refactor
+      const rawTree = harden({
+        [QCLASS]: 'error',
+        message,
+        name,
+        ...rest,
+      });
+      return decodeErrorFromCapData(rawTree);
+    };
+
+    const reviveFromSmallcaps = makeDecodeFromSmallcaps({
+      decodeRemotableFromSmallcaps,
+      decodePromiseFromSmallcaps,
+      decodeErrorFromSmallcaps,
+    });
+
+    const reviveFromCapData = makeDecodeFromCapData({
       decodeRemotableFromCapData: decodeRemotableOrPromiseFromCapData,
       decodePromiseFromCapData: decodeRemotableOrPromiseFromCapData,
       decodeErrorFromCapData,
     });
-    return fullRevive;
+
+    return harden({
+      reviveFromCapData,
+      reviveFromSmallcaps,
+    });
   };
+
   /**
    * @type {Unserialize<Slot>}
    */
   const unserialize = data => {
+    const { body, slots } = data;
     assert.typeof(
-      data.body,
+      body,
       'string',
-      X`unserialize() given non-capdata (.body is ${data.body}, not string)`,
+      X`unserialize() given non-capdata (.body is ${body}, not string)`,
     );
     (isArray(data.slots)) ||
       assert.fail(X`unserialize() given non-capdata (.slots are not Array)`);
-    const rawTree = harden(JSON.parse(data.body));
-    const fullRevive = makeFullRevive(data.slots);
-    const result = harden(fullRevive(rawTree));
+    const { reviveFromCapData, reviveFromSmallcaps } = makeFullRevive(slots);
+    let result;
+    // JSON cannot begin with a '#', so this is an unambiguous signal.
+    if (body.startsWith('#')) {
+      const smallcapsBody = body.slice(1);
+      const encoding = harden(JSON.parse(smallcapsBody));
```

Comment:
There's probably a future optimization that skips this `harden` via e.g. a revive option. It's currently necessary for the deep `passStyleOf` calls, but things might be noticeably faster if we didn't walk the data structure twice (once to recursively freeze and once to actually decode data).

Reply:
Skipping this one.